staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -117,28 +117,38 @@ type listenAddress struct {
 	failureMode string
 }
 
+// parseAddresses parses, validates addresses
+// Keeps the order from addressesToParse, but removes duplicates
 func parseAddresses(addressesToParse []string) ([]listenAddress, error) {
-	var addresses []listenAddress
+	var keyOrder []string
 	parsed := make(map[string]listenAddress)
 	for _, address := range addressesToParse {
 		if address == "localhost" {
-			ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
-			parsed[ip.address] = ip
-			ip = listenAddress{address: "::1", protocol: "tcp6", failureMode: "all"}
-			parsed[ip.address] = ip
+			if _, exists := parsed["127.0.0.1"]; !exists {
+				ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
+				parsed[ip.address] = ip
+				keyOrder = append(keyOrder, ip.address)
+			}
+			if _, exists := parsed["::1"]; !exists {
+				ip := listenAddress{address: "::1", protocol: "tcp6", failureMode: "all"}
+				parsed[ip.address] = ip
+				keyOrder = append(keyOrder, ip.address)
+			}
 		} else if net.ParseIP(address).To4() != nil {
 			parsed[address] = listenAddress{address: address, protocol: "tcp4", failureMode: "any"}
+			keyOrder = append(keyOrder, address)
 		} else if net.ParseIP(address) != nil {
 			parsed[address] = listenAddress{address: address, protocol: "tcp6", failureMode: "any"}
+			keyOrder = append(keyOrder, address)
 		} else {
 			return nil, fmt.Errorf("%s is not a valid IP", address)
 		}
 	}
-	addresses = make([]listenAddress, len(parsed))
-	id := 0
-	for _, v := range parsed {
-		addresses[id] = v
-		id++
+
+	parsedCount := len(parsed)
+	addresses := make([]listenAddress, parsedCount)
+	for id, mapKey := range keyOrder[:parsedCount] {
```

Comment:
this will truncate incorrectly with args like this:

localhost,127.0.0.1,10.10.10.1

parsedCount will be 3, and keyOrder will have "127.0.0.1" present twice

if you drop the keyOrder logic and revert changes from line 137 on, this will do what you want

Reply:
@liggitt oh, right... Do I understand correctly that we do not worry about order of the output (it will be different from time to time)?

I did add this `keyOrder` logic to always have the same output for the same input. It was previously causing issues in end-to-end tests, but now these tests do not rely on order. Just wondering if any of the users rely on it or not. I can't think of any useful use cases, so I'm happy to drop it.

Reply:
> Do I understand correctly that we do not worry about order of the output (it will be different from time to time)?

if you want to sort the addresses by IP before returning to get a stable order, that's fine

Reply:
Oh, yes. That makes more sense than trying to save the order specified in the input (in `addressesToParse`).

Thanks for the feedback. I'll update the PR on the weekend.